test/rateLimiters/slidingWindowLog.test.ts

```diff
@@ -0,0 +1,568 @@
+import 'ts-jest';
+import * as ioredis from 'ioredis';
+import { RateLimiterResponse, RedisLog } from '../../src/@types/rateLimit';
+import SlidingWindowLog from '../../src/rateLimiters/slidingWindowLog';
+
+// eslint-disable-next-line @typescript-eslint/no-var-requires
+const RedisMock = require('ioredis-mock');
+
+const WINDOW_SIZE = 1000;
+const CAPACITY = 10;
+
+let limiter: SlidingWindowLog;
+let client: ioredis.Redis;
+let timestamp: number;
+const user1 = '1';
+const user2 = '2';
+const user3 = '3';
+
+async function getLogFromClient(redisClient: ioredis.Redis, uuid: string): Promise<RedisLog> {
+    const res = await redisClient.get(uuid);
+    // if no uuid is found, return -1 for tokens and timestamp, which are both impossible
+    if (res === null) return [];
+    return JSON.parse(res);
+}
+
+async function setLogInClient(redisClient: ioredis.Redis, uuid: string, log: RedisLog) {
+    await redisClient.set(uuid, JSON.stringify(log));
+}
+
+/**
+ * Strategy
+ *
+ * Log and Redis updates
+ *  Doesn't exist
+ *    1. Request with complexity 0 => allowed.
+ *    2. Request with complexity < capacity => allowed
+ *    3. Request with complexity = capacity => allowed
+ *    4. Request with complexity > capacity => blocked
+ *  Empty
+ *    1. Request with complexity 0 => allowed.
+ *    2. Request with complexity < capacity => allowed
+ *    3. Request with complexity = capacity => allowed
+ *    4. Request with complexity > capacity => blocked
+ *  Contains active requests (still in window)
+ *    1. sum of requests = capacity => blocked
+ *    2. sum of request < capacity
+ *      1. current request complexity small enough => allowed
+ *      1. current request complexity remaining complexity  => allowed TODO:
+ *      2. current request complexity to big => blocked
+ *      3. current request complexity = 0 => allowed
+ *  Contains expired requests (no longer in the window)
+ *    1. Request with complexity 0 => allowed.
+ *    2. Request with complexity < capacity => allowed
+ *    3. Request with complexity = capacity => allowed
+ *    4. Request with complexity > capacity => blocked
+ *  Contains active and expired requests (both in and out of the window)
+ *    1. Request with complexity 0 => allowed.
+ *    2. Request with complexity < capacity => allowed
+ *    3. Request with complexity = remaining capacity => allowed
+ *    4. Request with complexity > capacity => blocked
+ *
+ * RateLimiter Functionality
```

Comment:
Should you have a test here for the `Retry After` header or leave that for the middle ware integration tests?